### ISSUE TITLE
Should time-based revalidation use a custom time string?

### ISSUE THREAD

**Issue by Asha20** [arctic-hen7/perseus#111]
I'm reading [this page of the documentation](https://arctic-hen7.github.io/perseus/en-US/docs/0.3.x/strategies/revalidation) and I'm looking at this part of the code in particular:

```rust
pub fn get_template<G: Html>() -> Template<G> {
    Template::new("timeisr")
        .template(time_page)
        // This page will revalidate every five seconds (to illustrate revalidation)
        .revalidate_after("5s".to_string())
        .incremental_generation()
        .build_state_fn(get_build_state)
        .build_paths_fn(get_build_paths)
}
```

I personally think it might be better to have `revalidate_after` take something like a `Duration`. That way we don't have to worry about string parsing which is a potentially fallible operation and can rely on Rust's type system instead. What do you think? I could open a PR if you agree.

**Comment by arctic-hen7** [comment]
Yeah, that's a much better idea. The only really nice thing about the time strings is that they can express complex durations like `1m2w4d3s` very efficiently.

I wonder if a good solution might be to have it take a `Duration`, but to then create a proc macro for parsing these strings for convenience, which would also solve the fallibility issues.

Also, making it take a `Duration` will unfortunately be a breaking change, so this could only be released in v0.4.0. A refactor to make that change simpler though and then maybe a commented out version to switch to then would be great though if you're up for it!

**Comment by Asha20** [comment]
Maybe you could even have it like this:

```rust
fn revalidate_after<I: Into<Duration>>(self, value: I) -> Self { ... }

struct DurationString { ... }
impl TryFrom<&str> for DurationString { ... }
impl From<DurationString> for Duration { ... }
```

And use it like so:

```rust
.revalidate_after(DurationString.try_from("5s").unwrap())
```

**Comment by arctic-hen7** [comment]
I considered that, but I was thinking compile-time errors might be cleaner (I can see most people just throwing a `.unwrap()` there, the easiest thing isn't the 'best' thing).

**Comment by Asha20** [comment]
Ah, I see. Yeah, having the compiler syntax-check the string for you could be pretty cool!

Unfortunately, I don't know much about writing proc macros so I couldn't implement that myself. I went ahead and implemented the approach I described in my previous comment, and then later down the line we could adapt the code by implementing a proc_macro which runs `DurationString::try_from` under the hood or something like that.

```rust
time_string!("foo") // Compile time error
time_string!("1w") // Valid DurationString
```

**Comment by arctic-hen7** [comment]
For reference, the linked PR has been merges manually due to conflicts, so this is now implemented!
